Readme.md
# OONI Explorer
[![This project is using Percy.io for visual regression testing.](https://percy.io/static/images/percy-badge.svg)](https://percy.io/407b7763/ooni-explorer)


For the live website see: https://explorer.ooni.org.

## Setup

We assume you have a working node.js development environment with yarn installed.

Then do:

```
yarn install
```

## Usage

To run the dev server do:

```
yarn run dev
```

To build the app:

```
yarn run build
```

To start the production server run:

```
yarn run start
```

We also provide a `Dockerfile` for easy deployment.


CHANGELOG.md
# Changelog

## [2.6.1] - 2021-05-20

### Added

- Enable anomaly search for Signal tests (#592)
- Link to github repo in footer (#592)

### Fixed

- Fixed broken feedback dialog (#592)


## [2.6.0] - 2021-05-17

### Added

- Enable anomaly search for whatsapp tests (#589)
- Show anomaly type in search results (#586)
- Move all environment variable definitions to .env.{production,development} files (#587)

### Changed

- Upgrade `next` to `v10.2.0`
- Automatically reformatted this file with `prettier` integration in VSCode :shrugs:

## [2.5.3] - 2021-04-27

### Added

- Show anomaly status in search results for middlebox tests

### Fixed

- Alignment of the floating feedback button (#583)

### Changed

- Control which tests can be filtered with anomaly and confirmed status

## [2.5.2] - 2021-04-22

### Added

- Summary text on RiseupVPN measurement pages

### Fixed

- Fix bug when selecting 'Any' in test name filter

## [2.5.1] - 2021-04-15

### Changed

- Group test names in search page sidebar into test groups
- Upgraded `next`, `ooni-components`, `react-icons` to latest versions.
- Upstream changes to component names of some icons from `react-icons`

## [2.5.0] - 2021-04-15

### Added

- Support for Signal tests (#575)

## [2.4.4] - 2021-03-18

### Changed

- Show the ok_count as the difference of confirmed and anomalies

### Fixed

- Fix temporarily API bug showing same anomaly and confirmed count
- Add experimental icon to MAT page

### Removed

- Remove accuracy disclaimer on whatsapp pages added in ca9526643b5fecb34c2794ea574bc83f7f4714d6
- Remove unused dependencies

## [2.1.1] - 2020-10-27

### Changed

- Date filters in search will default to a 30 day range unless manually changed.
  This is to align with the new default in the `/api/v1/measurements` API. (#512)

### Fixed

- Fix DNS Queries box by handling AAAA answer types correctly (#505)

## [2.1.0] - 2020-10-15

### Changed

- **Uses the new refactored API**. Incompatible with 2.0.x because the endpoints to fetch measurement data has changed. (#486)

### Fixed

- Fix link colors to use `blue7` from the theme (#473)
- Address warnings about using `next/link` for external links (#484)
- Ensure that measurment pages don't crash when missing test keys (#491)

### Added

- Responsive header and nav menu in countries page (#454)
- Links to test descriptions on measurement pages (#469)
- Social media sharing buttons on measurement and country pages. This was done by our awesome [GSoC 2020 student](https://medium.com/@kronaemmanuel/anxious-to-excitement-my-gsoc-story-begins-c83041900f35) @kronaemmanuel (#476)
- Dev tool to toggle parts of `test_keys` in measurement data (#491)

### Removed

- Remove tested since date on country pages (#483)

### Security

- Bump elliptic from 6.4.1 to 6.5.3 (#477)
- Bump markdown-to-jsx from 6.9.4 to 6.11.4 (#485)

## [2.0.10] - 2020-10-07

### Fixed

- Address the OONI Probe ASN Incident reported [here](https://ooni.org/post/2020-ooni-probe-asn-incident-report/) by making it harder to lookup and access measurements with AS0 (#495)

## [2.0.9] - 2020-05-09

### Fixed

- Country page meta tag now correctly shows the measurement count for the
  country

### Changed

- Search filter test case uses test name filter to test search function
  instead of domain name filter. Avoids timeouts when API is slow.

### Security

- Pinned version for `kind-of` to `^6.0.3` as per github security advisory

## [2.0.8] - 2020-05-07

### Added

- Support for ndt7 measurements (#456)
- Meta tags for country pages (#430)
- End-to-end tests with Cypress (#417)

### Fixed

- Missing overview charts in country page (#459)
- Country codes in URL for country pages are not case sensitive anymore (#449)
- Raw measurement data content overflow in smaller screens (#429)
- Handle cases where `test_keys.tcp_connect` is missing (#412)
- Handle cases where `test_keys.requests` is missing (#412)

### Changed

- `nextjs` dependecy upgraded to 9.3.2 (#448)
- Upgrade base docker image to `node:12` as required by `next@9.x.x`
- Use `blue9` color from theme for footer background (#411)
- Updated copy

### Removed

- Dropped usage of locale-data from `react-intl`; deprecated (in v3)

### Security

- Pinned version for `https-proxy-agent` to `^2.2.3` as per github security advisory

## [2.0.7] - 2020-02-24

### Added

- Circumvention test group containing the new psiphon and tor tests (#372, #393)
- Use `scores` key from `fastpath` to show endpoint status for WhatsApp test (#405)
- Added disclaimer for older measurements (ca95266)

### Fixed

- Fixed alignment of search result rows for dns_consistency tests (#387)
- Fixed bug in selecting the correct date range in search (#397)
- Made usage of blocking reason test_key more robust (#401)
- Fixed spin loader crash by upgrading it to styled-components v4 (#408)
- Fix for bug in search results with `#` in `input` field (#409)

### Changed

- Changed ooni.io links to ooni.org (#372)
- Upgraded styled-components to v4 (#398)
- Use `country-util` package instead of API call to fetch countries list (#394)
- Switch back to using `anomaly` flag from API instead of determining from test_keys(#404)

## [2.0.6] - 2020-01-24

### Fixed

- Fixed bug in rendering measurements with misaligned anomaly flag (#374)
- Fixed error message copy for search page failures (#370)

## [2.0.5] - 2020-01-09

### Added

- Added regex to allow search by IP address in the domain field (#367)
- Added matomo based analytics (#369)

### Fixed

- Fixed bug where duplicate API requests were made for search queries (#353)
- Fixed bug in rendering DNS queries on WebConnectivity measurement pages (#363)
- Fixed 'until today' logic in search to include today's measurement (#364)

## [2.0.4] - 2019-11-27

### Added

- Added this changelog
- Added 'UTC' label to measurement timestamps (#328)
- Added explorer version in footer (#329)
- Added link to fastpath event detector feed (#342)

### Changed


pages/index.js
/* global process */
import React from 'react'
import PropTypes from 'prop-types'
import Head from 'next/head'
import NLink from 'next/link'
import Router from 'next/router'
import FormattedMarkdown from '../components/FormattedMarkdown'
import styled from 'styled-components'
import axios from 'axios'
import { FormattedMessage } from 'react-intl'
import {
  Link,
  Flex,
  Box,
  Container,
  Button,
  Heading,
  Text
} from 'ooni-components'

import Layout from '../components/Layout'
import NavBar from '../components/NavBar'
import { toCompactNumberUnit } from '../utils'
import HighlightSection from '../components/landing/HighlightsSection'
import highlightContent from '../components/landing/highlights.json'
import CoverageChart from '../components/landing/Stats'

const HeroUnit = styled.div`
  background: linear-gradient(
    319.33deg,
    ${props => props.theme.colors.blue9} 39.35%,
    ${props => props.theme.colors.base} 82.69%),
    ${props => props.theme.colors.base};
  background-size: contain;
  background-repeat: no-repeat;
  background-position: center;
  padding-bottom: 48px;
  padding-top: 16px;
`

const ExploreButton = styled(Button)`
  font-weight: normal;
  color: white;
  border: 2px solid white;
  border-radius: 48px;
  height: 60px;
  cursor: pointer; /* until added to Button in ooni-components */

  &:hover, &:focus {
    background-color: white;
    color: ${props => props.theme.colors.blue5};
  }

  &:active {
    background-color: white;
    color: ${props => props.theme.colors.blue5};
    box-shadow: 0 0 0 0.2rem ${props => props.theme.colors.blue4};
  }
`

const StatsContainer = styled(Flex)`
  border-radius: 16px;
  background-color: #ffffff;
`

const StyledLabel = styled.div`
  color: ${props => props.theme.colors.gray7};
`

const StyledStatsItem = styled(Box)`
  text-align: center;
`

const StatsItem = ({label, unit, value }) => (
  <StyledStatsItem color='blue9' width={[1/3]} p={3}>
    <Text fontSize={[42, 48]} fontWeight={300}>
      {value}
      <Text as='span' fontSize={32}>{unit}</Text>
    </Text>
    <StyledLabel>
      {label}
    </StyledLabel>
  </StyledStatsItem>
)

StatsItem.propTypes = {
  label: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.string
  ]),
  unit: PropTypes.string,
  value: PropTypes.number
}

const FeatureRow = styled(Flex)`

`

FeatureRow.defaultProps = {
  flexWrap: 'wrap',
  alignItems: 'center',
  justifyContent: 'center',
  py: 4
}

const FeatureBox = styled(Box)`
  font-size: 20px;
  line-height: 1.5;
`

FeatureBox.defaultProps = {
  width: [1, 1/2],
  lineHeight: 1.5,
}

const FeatureBoxTitle = styled(Text)`
`
FeatureBoxTitle.defaultProps = {
  textAlign: ['center', 'left'],
  color: 'blue9',
  fontSize: 24,
  fontWeight: 600,
  mb: 2
}

const ImgBox = styled.img`
  width: 100%;
`

const StyledContainer = styled(Container)`
  background-image: url('/static/images/world-dots.svg');
  background-repeat: no-repeat;
  background-position: center;
`

export default class LandingPage extends React.Component {

  static async getInitialProps () {
    const client = axios.create({baseURL: process.env.NEXT_PUBLIC_MEASUREMENTS_URL}) // eslint-disable-line
    const result = await client.get('/api/_/global_overview')
    return {
      measurementCount: result.data.measurement_count,
      asnCount: result.data.network_count,
      countryCount: result.data.country_count
    }
  }

  constructor(props) {
    super(props)
  }

  render () {
    let {
      measurementCount,
      asnCount,
      countryCount
    } = this.props

    measurementCount = toCompactNumberUnit(measurementCount)
    asnCount = toCompactNumberUnit(asnCount)

    return (
      <Layout>
        <Head>
          <title>OONI Explorer</title>
        </Head>
        <HeroUnit>
          <NavBar color='transparent' />
          <StyledContainer py={[0, 120]} my={[0, 90]}>
            <Text textAlign='center'>
              <Heading h={1}>
                <Text fontSize={[32, 64]} color='#ffffff'><FormattedMessage id='Home.Banner.Title.UncoverEvidence' /></Text>
              </Heading>
              <Text fontSize={[18, 24]} color='blue1'><FormattedMessage id='Home.Banner.Subtitle.ExploreCensorshipEvents' /></Text>
              <ExploreButton mt={48} px={5} hollow fontSize={24} onClick={() => (
                Router.push('/countries')
              )}>
                <FormattedMessage id='Home.Banner.Button.Explore' />
              </ExploreButton>
            </Text>
          </StyledContainer>
        </HeroUnit>
        <Container>
          <StatsContainer px={[0, 32]} py={16} mx={[0, '25%']} mt={[0, -120]} mb={48} flexWrap='wrap'>
            <StatsItem
              label={<FormattedMessage id='Home.Banner.Stats.Measurements' />}
              unit={measurementCount.unit}
              value={measurementCount.value}
            />
            <StatsItem
              label={<FormattedMessage id='Home.Banner.Stats.Countries' />}
              value={countryCount}
            />
            <StatsItem
              label={<FormattedMessage id='Home.Banner.Stats.Networks' />}
              unit={asnCount.unit}
              value={asnCount.value}
            />
          </StatsContainer>

          {/* Intro text about Explorer */}
          <Flex justifyContent='center' my={4}>
            <Box width={[1, 2/3]}>
              <Text fontSize={20} lineHeight={1.5}>
                <FormattedMarkdown id='Home.About.SummaryText' />
              </Text>
            </Box>
          </Flex>

          {/* Websites & Apps */}
          <FeatureRow>
            <FeatureBox>
              <ImgBox src='/static/images/websites-apps.png' alt='Websites and Apps' />
            </FeatureBox>
            <FeatureBox color='gray7'>
              <FeatureBoxTitle>
                <FormattedMessage id='Home.Websites&Apps.Title' />
              </FeatureBoxTitle>
              <FormattedMessage id='Home.Websites&Apps.SummaryText' />
            </FeatureBox>
          </FeatureRow>
          {/* Search & Filter */}
          {/* Arrange in {[img, para], [img, para], [img, para]} pattern on smaller screens */}
          <FeatureRow flexDirection={['column-reverse', 'row']}>
            <FeatureBox color='gray7'>
              <FeatureBoxTitle>
                <FormattedMessage id='Home.Search&Filter.Title' />
              </FeatureBoxTitle>
              <FormattedMessage id='Home.Search&Filter.SummaryText' />
            </FeatureBox>
            <FeatureBox>
              <ImgBox src='/static/images/search.png' alt='Search and Filter' />
            </FeatureBox>
          </FeatureRow>
          {/* Network Properties */}
          <FeatureRow>
            <FeatureBox>
              <ImgBox src='/static/images/network-performance.png' alt='Network Properties' />
            </FeatureBox>
            <FeatureBox color='gray7'>
              <FeatureBoxTitle>
                <FormattedMessage id='Home.NetworkProperties.Title' />
              </FeatureBoxTitle>
              <FormattedMessage id='Home.NetworkProperties.SummaryText' />
            </FeatureBox>
          </FeatureRow>
          {/* Measurement Statistics */}
          <Container mb={5}>
            <Flex justifyContent='center' my={3}>
              <Heading h={2} color='blue7'>
                <FormattedMessage id={'Home.MonthlyStats.Title'} />
              </Heading>
            </Flex>
            <CoverageChart />
          </Container>
          {/* Highlights */}
          <Container>
            <Flex flexWrap='wrap' justifyContent='center' my={3}>
              <Heading h={2} color='blue7'>
                <a id='highlights'>
                  <FormattedMessage id={'Home.Highlights.Title'} />
                </a>
              </Heading>
            </Flex>
            <Flex flexWrap='wrap' justifyContent='center'>
              <Box px={[0, 4]} my={3}>
                <Text textAlign='center' fontSize={26}>
                  <FormattedMarkdown
                    id='Home.Highlights.Description'
                  />
                </Text>
              </Box>
            </Flex>

            {/* Political Events */}
            <HighlightSection
              title={<FormattedMessage id='Home.Highlights.Political' />}
              description={<FormattedMessage id='Home.Highlights.Political.Description' />}
              highlights={highlightContent.political}
            />
            {/* Media */}
            <HighlightSection
              title={<FormattedMessage id='Home.Highlights.Media' />}
              description={<FormattedMessage id='Home.Highlights.Media.Description' />}
              highlights={highlightContent.media}
            />
            {/* LGBTQI sites */}
            <HighlightSection
              title={<FormattedMessage id='Home.Highlights.LGBTQI' />}
              description={<FormattedMessage id='Home.Highlights.LGBTQI.Description' />}
              highlights={highlightContent.lgbtqi}
            />
            {/* Censorship changes */}
            <HighlightSection
              title={<FormattedMessage id='Home.Highlights.Changes' />}
              description={<FormattedMessage id='Home.Highlights.Changes.Description' />}
              highlights={highlightContent.changes}
            />
            <Box my={3}>
              <Text fontSize={20}>We encourage you to <NLink href='/search' passHref><Link color='blue7'>explore OONI measurements</Link></NLink> to find more highlights!</Text>
            </Box>
          </Container>
        </Container>
      </Layout>
    )
  }
}

LandingPage.propTypes = {
  countryCount: PropTypes.number,
  asnCount: PropTypes.number,
  measurementCount: PropTypes.number
}
